infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStorageProvisioner.java

```diff
@@ -333,19 +336,37 @@ private void createAsyncStoragePodIfNotExist(
     PodSpec podSpec =
         podSpecBuilder.withContainers(container).withVolumes(storageVolume, sshKeyVolume).build();
 
-    Pod pod =
-        new PodBuilder()
-            .withApiVersion("v1")
-            .withKind("Pod")
-            .withNewMetadata()
-            .withName(ASYNC_STORAGE)
+    ObjectMetaBuilder metaBuilder = new ObjectMetaBuilder();
+    ObjectMeta meta =
+        metaBuilder
+            .withLabels(
+                of(
+                    "app",
+                    ASYNC_STORAGE,
+                    CHE_USER_ID_LABEL,
+                    userId,
+                    CHE_DEPLOYMENT_NAME_LABEL,
+                    ASYNC_STORAGE))
             .withNamespace(namespace)
-            .withLabels(of("app", ASYNC_STORAGE, CHE_USER_ID_LABEL, userId))
-            .endMetadata()
+            .withName(ASYNC_STORAGE)
+            .build();
+
+    Deployment deployment =
+        new DeploymentBuilder()
+            .withMetadata(meta)
+            .withNewSpec()
+            .withNewSelector()
+            .withMatchLabels(meta.getLabels())
+            .endSelector()
+            .withReplicas(1)
+            .withNewTemplate()
+            .withMetadata(meta)
             .withSpec(podSpec)
+            .endTemplate()
+            .endSpec()
             .build();
 
-    k8sClient.pods().inNamespace(namespace).create(pod);
+    k8sClient.apps().deployments().inNamespace(namespace).create(deployment);
```

Comment:
is it ok to create deployment here directly in the Provisioner? Should it be created later in `KubernetesInternalRuntime` together with all other objects?

Reply:
It Async Storage deployment will be used for several workspaces, so i think it's ok to create deployment here

Reply:
> It Async Storage deployment will be used for several workspaces, so i think it's ok to create deployment here

The whole point of provisioners filling up the environment which then "realized" by some runtime is that we are runtime independent (at least to an extent). We can debate about usefulness of that when we only have kubernetes-based runtimes, but we should start changing the flow without a broader discussion (of which I am unaware).

I recommend following the common approach and kudos to @sparkoo for spotting this.

Reply:
So in this case we will shutdown AsyncStorage service with workspace configured to use Async Storage feature

Reply:
@metlos @sparkoo 
 I mean that we don't need to stop `AsyncStorage` immediately on workspace stop, that's why I don't add it to the `KubernetesEnvironment`. We need to introduce ability to not stop marked Pods/Deployments on workspace stop.
As I know we don't have such ability at the moment.

Reply:
@metlos @sparkoo if we want to add this deployment to `KubernetesInternalRuntime` we need to mark it somehow as `unstoppable` on the workspace stop. What about some labels or attributes?

Reply:
what about moving all this from provisioning phase into the `KubernetesInternalRuntime`? It is not technically workspace provisioning, it's something on top of it. It's not bound to single workspace and it has different lifecycle.

Reply:
+0.5. @metlos wdyt about @sparkoo idea to move this from provisioning?

Reply:
@skabashnyuk @sparkoo @metlos 
Ok, guys probably you are right, as you have much more knowledge and experience in this area.  Your proposals look reasonable and valuable. But looks like it's not related to changes in this PR. On this PR I've just replace Pod with Deployment. So my proposal create separate issue on such improvement, discus more detail and  found the best solution for this. WDYT?

Reply:
I agree. We should've raise this in original PR. This PR does not make things worse so it does not make sense to block it at this point, because of this refactoring. It make sense to me to keep it to separate issue.